src/box/lua/config/applier/box_cfg.lua

```diff
@@ -32,24 +19,50 @@ local function peer_uris(configdata)
         -- An anonymous replica can be an upstream for another
         -- anonymous replica, but only non-anonymous peers are set
         -- as upstreams by default.
-        --
-        -- A user may configure a custom data flow using
-        -- `replication.peers` option.
         if not is_anon then
-            local uri = instance_config:instance_uri(iconfig_def, 'peer')
-            if uri == nil then
-                log.info('%s: instance %q has no iproto.advertise.peer or ' ..
-                    'iproto.listen URI suitable to create a client socket',
-                    err_msg_prefix, peer_name)
-            end
-            if uri ~= nil and peer_name ~= names.instance_name then
-                has_upstream = true
-            end
-            table.insert(uris, uri)
+            table.insert(peers, peer_name)
+        end
+    end
+    --
+    -- Replicaset should have at least one non-anonymous replica and it was
+    -- already validated in validate_anon().
+    --
+    assert(#peers > 0)
+    return peers
+end
+
+local function peer_uris(configdata)
+    local peers = instance_config:get(configdata._iconfig_def,
+        'replication.peers')
+    if peers == nil then
+        --
+        -- No need to set box.cfg.replication if there is only one instance in
+        -- the replicaset.
+        --
+        if #configdata:peers() == 1 then
+            return box.NULL
+        end
+        peers = auto_upstreams(configdata)
+    end
+
+    local names = configdata:names()
+    local err_msg_prefix = ('replication.peers construction for instance %q ' ..
+        'of replicaset %q of group %q'):format(names.instance_name,
+        names.replicaset_name, names.group_name)
+
+    local uris = {}
+    for _, peer_name in ipairs(peers) do
+        local iconfig_def = configdata._peers[peer_name].iconfig_def
+        local uri = instance_config:instance_uri(iconfig_def, 'peer')
+        if uri == nil then
+            log.info('%s: instance %q has no iproto.advertise.peer or ' ..
+                'iproto.listen URI suitable to create a client socket',
+                err_msg_prefix, peer_name)
         end
+        table.insert(uris, uri)
     end
 
-    if not has_upstream then
+    if next(uris) == nil then
         error(('%s: no suitable peer URIs found'):format(err_msg_prefix), 0)
     end
```

Comment:
I need to think about this check a bit more.

The logic behind the former check is the following: it is OK while any node has at least one upstream (except itself). The check is not performed for a single-node replicaset. It is also not performed for a user provided list of URIs.

It has the following problem.

* It doesn't allow to bootstrap a replicaset with one non-anonymous instance and all the other instances as anonymous replicas.

Next, if we're going to use instance names instead of URIs, we can verify a user-provided list of instances.

The 'has at least one upstream' check is insufficient on a replicaset bootstrap (it is still possible to bootstrap two separate replicasets) and, at the same time, may be too tight during a maintenance of an existing replicaset (we may want to temporary stop replication to the given instance).